### PR TITLE
Added `--only-funcptrs` feature to spray comand

### PR DESCRIPTION
**Pull request by bog2n**
Added --only-funcptrs flag that sprays only values in memory that are pointers to executable area in memory (presumably function pointers)

### CODE REVIEW

**Comment by disconnect3d** [comment]
This will close https://github.com/pwndbg/pwndbg/issues/1778

File: pwndbg/commands/spray.py [hunk 1]
```diff
@@ -21,10 +22,11 @@
     type=str,
     required=False,
 )
+parser.add_argument("-x", "--only-funcptrs", help="Spray functionpointers", action="store_true")
```
Comment (disconnect3d) [inline pwndbg/commands/spray.py]:
```suggestion
parser.add_argument("-x", "--only-funcptrs", help="Spray only addresses whose values points to executable pages", action="store_true")
```

File: pwndbg/commands/spray.py [hunk 2]
```diff
@@ -52,7 +54,26 @@ def spray(addr, length, value) -> None:
...
 
-    try:
-        pwndbg.gdblib.memory.write(addr, value_bytes)
-    except gdb.MemoryError as e:
-        print(M.error(e))
+    if only_funcptrs:
+        try:
+            mem = pwndbg.gdblib.memory.read(addr, length)
+        except gdb.MemoryError as e:
+            print(M.error(e))
```
Comment (disconnect3d) [inline pwndbg/commands/spray.py:66]:
```suggestion
            print(M.error(e))
            return
```

File: pwndbg/commands/spray.py [hunk 3]
```diff
@@ -52,7 +54,27 @@ def spray(addr, length, value) -> None:
...
+            return
+
+        addresses_written = 0
+        ptrsize = pwndbg.gdblib.arch.ptrsize
```
Comment (disconnect3d) [inline pwndbg/commands/spray.py:70]: Can we move this to the begining of the whole function? There is a `pwndbg.gdblib.arch.ptrsize` used in line 53 that could re-use this :P

File: pwndbg/commands/spray.py [hunk 4]
```diff
@@ -52,7 +54,27 @@ def spray(addr, length, value) -> None:
...
+        for i in range(0, len(mem) - length % ptrsize, ptrsize):
+            ptr_candidate = struct.unpack(pwndbg.gdblib.arch.fmt, mem[i : i + ptrsize])[0]
+            page = pwndbg.gdblib.vmmap.find(ptr_candidate)
+            if page is not None and page.execute:
+                try:
+                    pwndbg.gdblib.memory.write(addr + i, value_bytes[i : i + ptrsize])
+                    addresses_written += 1
+                except gdb.MemoryError as e:
+                    print(M.error(e))
```
Comment (disconnect3d) [inline pwndbg/commands/spray.py:79]:
```suggestion
                    print(M.error(e))
                    return
```

File: pwndbg/commands/spray.py [hunk 5]
```diff
@@ -52,7 +59,28 @@ def spray(addr, length, value) -> None:
...
+        try:
+            mem = pwndbg.gdblib.memory.read(addr, length)
+        except gdb.MemoryError as e:
+            print(M.error(e))
+            return
+
+        addresses_written = 0
+        ptrsize = pwndbg.gdblib.arch.ptrsize
+        for i in range(0, len(mem) - length % ptrsize, ptrsize):
+            ptr_candidate = struct.unpack(pwndbg.gdblib.arch.fmt, mem[i : i + ptrsize])[0]
```
Comment (gsingh93) [inline pwndbg/commands/spray.py:72]: You can use [`Arch.unpack`](https://github.com/bog2n/pwndbg/blob/e02ff210eb1b8bc8e941a9fbed57351833df8316/pwndbg/lib/arch.py#L37) here.

File: pwndbg/commands/spray.py [hunk 6]
```diff
@@ -52,7 +59,28 @@ def spray(addr, length, value) -> None:
...
+    if only_funcptrs:
+        try:
+            mem = pwndbg.gdblib.memory.read(addr, length)
+        except gdb.MemoryError as e:
+            print(M.error(e))
+            return
+
+        addresses_written = 0
+        ptrsize = pwndbg.gdblib.arch.ptrsize
+        for i in range(0, len(mem) - length % ptrsize, ptrsize):
```
Comment (gsingh93) [inline pwndbg/commands/spray.py:71]: nit: I'd recommend parenthesis here just for clarity: `len(mem) - (length % ptrsize)`

File: pwndbg/commands/spray.py [hunk 7]
```diff
@@ -52,7 +59,28 @@ def spray(addr, length, value) -> None:
...
+                    addresses_written += 1
+                except gdb.MemoryError as e:
+                    print(M.error(e))
+                    return
+        print(M.notice(f"Overwritten {addresses_written} function pointers"))
+    else:
+        try:
+            pwndbg.gdblib.memory.write(addr, value_bytes)
+        except gdb.MemoryError as e:
+            print(M.error(e))
```
Comment (gsingh93) [inline pwndbg/commands/spray.py:86]: nit: I think we can probably combine the various `gdb.MemoryError` exception handlers into just one or two?

Comment (disconnect3d) [reply pwndbg/commands/spray.py:86]: Fixed in https://github.com/pwndbg/pwndbg/pull/1819

File: pwndbg/commands/spray.py (hunk 6 above)
Comment (disconnect3d) [reply pwndbg/commands/spray.py:71]: Fixed in https://github.com/pwndbg/pwndbg/pull/1819

File: pwndbg/commands/spray.py (hunk 5 above)
Comment (disconnect3d) [reply pwndbg/commands/spray.py:72]: Fixed in https://github.com/pwndbg/pwndbg/pull/1819

**Pull request merged**